### ISSUE TITLE
Next.js 13 support

### ISSUE THREAD

**Issue by thorwebdev** [supabase/auth-helpers#341]
Next.js 13 drastically changes how data is being fetched and will require fundamental changes: https://beta.nextjs.org/docs/upgrade-guide 

E.g. currently it isn't clear how to access what used to be `ctx.res` in getServerSideProps. This part is missing in the upgrade guide: https://beta.nextjs.org/docs/upgrade-guide#accessing-request-object

**Comment by Mar0xy** [comment]
> E.g. currently it isn't clear how to access what used to be ctx.res in getServerSideProps

Seems like context was generally replaced in favor of the headers and cookies module as it stands now you can only read from both but according to Next.js they are planning on implementing a set function for cookies so you can set cookies.

"The Next.js team at Vercel is working on adding the ability to set cookies in addition to the cookies function. For now, if you need to set cookies, you can use Middleware."

**Comment by Murkrage** [comment]
I'm going to spend some time today to see if I can figure out a relatively straightforward way to get this working without too many drastic changes.

**Comment by jensen** [comment]
I spent some time yesterday and a bit today. Was able to use the existing `withMiddlewareAuth` function directly.

```javascript
export const middleware = withMiddlewareAuth({
  redirectTo: "/login",
});
```

In order to get access to a supabase client that is authenticated on the server, I am doing something like this:

```javascript
import { cookies } from "next/headers";
import { createClient } from "@supabase/supabase-js";

export const createSupabaseClient = () => {
  if (!process.env.NEXT_PUBLIC_SUPABASE_URL) {
    throw new Error(`Must define process.env.NEXT_PUBLIC_SUPABASE_URL`);
  }

  if (!process.env.NEXT_PUBLIC_SUPABASE_ANON_KEY) {
    throw new Error(`Must define process.env.NEXT_PUBLIC_SUPABASE_URL`);
  }

  const cookie = JSON.parse(cookies().get("supabase-auth-token") || "{}");

  if (cookie?.access_token) {
    return createClient(
      process.env.NEXT_PUBLIC_SUPABASE_URL,
      process.env.NEXT_PUBLIC_SUPABASE_ANON_KEY,
      {
        auth: {
          persistSession: false,
        },
        global: {
          headers: {
            Authorization: `Bearer ${cookie.access_token}`,
          },
        },
      }
    );
  }

  return createClient(
    process.env.NEXT_PUBLIC_SUPABASE_URL,
    process.env.NEXT_PUBLIC_SUPABASE_ANON_KEY,
    {
      auth: {
        persistSession: false,
      },
    }
  );
};
```

Which I'm calling in my server component. 

```javascript
import { createSupabaseClient } from "services/supabase";

const getCustomers = async () => {
  const supabase = createSupabaseClient();

  return (await supabase.from("customers").select()).data;
};

const Customers = async () => {
  const customers = await getCustomers();

  return <pre>{JSON.stringify(customers, null, 2)}</pre>;
};

export default Customers;
```

I don't expect this to be a long term solution, and it doesn't have the ability to set the cookies yet, but for authenticated requests I don't need to at this point. Just wanted to give a quick progress update. I'll be posting the repo once I do a bit more testing.

**Comment by thorwebdev** [comment]
Thanks @jensen I actually think pairing it with middleware is the right approach because the middleware can modify the response to refresh the session and update the cookie if needed: https://github.com/supabase/auth-helpers/blob/main/packages/nextjs/src/middleware/withMiddlewareAuth.ts#L65-L72

**Comment by jensen** [comment]
Yeah, I was thinking the same thing. The only issue I'm finding with that at this moment is passing that `supabase` instance to the server component isn't straight forward. Right now we have access to cookies and headers, but not a context like object passed from middleware.

I'd like to look at the middleware in more detail. We need access to the request, and the response in order to create the storage callbacks, that part is reasonable. The part that I am unsure of is how we pass a `supabase` instance to a server component. We can access `cookies` and `headers` at this point but it seems like no complex types.

Would love to be wrong on this.

**Comment by jensen** [comment]
> Has anyone worked out how to use withPageAuth yet? or is this still a work in progress.
> 
> Also how long util there is a supported solution @supabase

The withPageAuth API likely won't be as useful for server components as it is for the existing `getServerSideProps` API. I imagine a new hook could be used instead. Maybe named `usePageAuth`, but currently we cannot write the redirect headers from a server component.

We can currently only read cookies and headers from the request. This is documented by the vercel team at https://beta.nextjs.org/docs/api-reference/cookies and https://beta.nextjs.org/docs/api-reference/headers.

**Comment by Murkrage** [comment]
I agree with @jensen that a new hook is probably the way to go. Looking at the [redirect example](https://beta.nextjs.org/docs/api-reference/redirect) on the new Next docs it seems like something along the following lines is gonna suffice:

```js
import { redirect } from 'next/navigation';
import { usePageAuth } from 'supabase/auth-helpers'; // This one would be new

export default async function PageUsingAuth() {
  const isAuthorized = usePageAuth();
  if (!isAuthorized) {
    redirect('https://...');
  }

  // ...
}
```

This would be done on a page level and then the child components are protected by default. I believe then using middleware to handle refreshing the session and cookie would be enough for this all to work. To me, this is a much nicer API than using `withPageAuth()`.

**Comment by jensen** [comment]
Just a heads up for anyone working on Next 13. I found out the hard way that leaving out a `pages` folder will prevent `middleware.ts` from loading. I had to add a `pages/.keep` file for now, and it works. I was going to add an issue, but found this one https://github.com/vercel/next.js/issues/41995 was already submitted.

**Comment by rodzerjohn** [comment]
Can anyone provide with example how to implement `SessionContextProvider` with new `app` page structure?

**Comment by jensen** [comment]
> Can anyone provide with example how to implement `SessionContextProvider` with new `app` page structure?

I posted an example repo in a message above.

**Comment by Murkrage** [comment]
> You could also make a version that works on the layout that way you can have a whole subset without the middleware

Currently the middleware is needed because there is no other way to modify the response of a request.

**Comment by The-Code-Monkey** [comment]
> Can anyone provide with example how to implement `SessionContextProvider` with new `app` page structure?

You have to implement it like normal the only thing thats different is that you have to put 'use client' at the top of the page that uses that component. as server components cannot use context. Which for me breaks every page as i use a Theme Context with styled components.

**Comment by The-Code-Monkey** [comment]
> > You could also make a version that works on the layout that way you can have a whole subset without the middleware
> 
> Currently the middleware is needed because there is no other way to modify the response of a request.

Is there an example of how to use the middleware with Next 13 and auth-helpers anywhere @Murkrage

**Comment by kentare** [comment]
> > > You could also make a version that works on the layout that way you can have a whole subset without the middleware
> > 
> > 
> > Currently the middleware is needed because there is no other way to modify the response of a request.
> 
> Is there an example of how to use the middleware with Next 13 and auth-helpers anywhere @Murkrage

@jensen provided one that works fine in my project: https://github.com/supabase/auth-helpers/issues/341#issuecomment-1292658432 

Code: https://github.com/jensen/supabase-nextjs/blob/main/middleware.ts 
```typescript
// middleware.ts
import { withMiddlewareAuth } from "@supabase/auth-helpers-nextjs";

export const middleware = withMiddlewareAuth({
  redirectTo: "/login",
});

export const config = {
  matcher: "/dashboard/:path*",
};

**Comment by The-Code-Monkey** [comment]
I have done this in mine. The app redirects to the login page but then when I fill in the form it submits the form the html way even though I've got e.preventDefault on it. If I move the login page back to pages it works but in app it doesn't. Super strange

**Comment by kentare** [comment]
If it is an existing project, have you followed the upgrade guide? https://beta.nextjs.org/docs/upgrade-guide

**Comment by The-Code-Monkey** [comment]
Yeah i've followed it all my pages and layouts are set up as 'use client' so nothing should have changed.

**Comment by The-Code-Monkey** [comment]
@kentare are you able to look and see if im doing anything wrong with my repo https://github.com/The-Code-Monkey/PaperCMS

**Comment by nicolazj** [comment]
https://github.com/nicolazj/nextjs13 here's a working example with nextjs13

**Comment by Mar0xy** [comment]
If this [PR](https://github.com/vercel/next.js/pull/42388) gets merged and put into canary for the next version of next.js I think it would then enable setting cookies everywhere.

**Comment by Bruce9000** [comment]
> https://github.com/nicolazj/nextjs13 here's a working example with nextjs13

I couldn't get it working :(

I cloned the repo, did an npm install, put in an .env.local with my supabase url and anon_key, started the dev server. 
Entered my username and password, hit sign in, was redirected to the Account page but the server shows "null from server" and the browser shows an alert with "Error loading user data!".
Am I missing a step somewhere for this to work?

**Comment by xsmithdev** [comment]
> https://github.com/nicolazj/nextjs13 here's a working example with nextjs13

@nicolazj I had success at first, but when moving from my Macbook to my Windows machine I keep getting "Cannot read properties of undefined (reading 'Provider')". Not sure if this is a windows issue however

**Comment by xsmithdev** [comment]
From what I can tell, using supabase/ui fails with Next 13

**Comment by hjaber** [comment]
Now that react & [next](https://beta.nextjs.org/docs/data-fetching/caching#per-request-caching) can cache fetched data, would it be possible to remove the context provider when using Nextjs 13 since the session can be natively cached? Is the context provider only caching the session?

This currently doesn't work, but something to the effect of: 

```
// utils/getSession.tsx
import { cache } from "react";
import { createServerSupabaseClient } from "@supabase/auth-helpers-nextjs";
import "server-only";

export const getSession = cache(async () => {
  const supabaseServerClient = createServerSupabaseClient();
  const { data: session } = await supabaseServerClient.auth.getSession();
  return session;
});
```
```
// app/user/page.tsx
import { getSession } from "@utils/getSession";

export default async function Page() {
  const session = await getSession();
  // ...
}
```

**Comment by thorwebdev** [comment]
With version `v0.5.2` we've added a `createServerComponentSupabaseClient` which can be used server components to get an authenticated Supabase client. [This is the PR](https://github.com/supabase/auth-helpers/pull/380) which includes an [example](https://github.com/supabase/auth-helpers/blob/main/examples/nextjs-server-components/app/required-session/page.tsx). Note that this needs to be used in combination with [middleware](https://github.com/supabase/auth-helpers/blob/main/examples/nextjs-server-components/middleware.tsx) as Next.js does not yet allow setting headers in server components.

Docs to follow soon.

**Comment by rodzerjohn** [comment]
How do i call inner nextjs api with session? `session` object in middleware returns `null`.

**Comment by yinkakun** [comment]
> How do i call inner nextjs api with session? `session` object in middleware returns `null`.

Same for me too.

**Comment by walton-alex** [comment]
Is there an ETA on the docs at all?

**Comment by dijonmusters** [comment]
> How do i call inner nextjs api with session? `session` object in middleware returns `null`.

Very strange. This should only happen if you are not signed in.

If you make your middleware something like this:

```
import { createMiddlewareSupabaseClient } from "@supabase/auth-helpers-nextjs";
import { NextResponse } from "next/server";
import type { NextRequest } from "next/server";

export async function middleware(req: NextRequest) {
  const res = NextResponse.next();
  const supabase = createMiddlewareSupabaseClient({ req, res });

  const {
    data: { session },
  } = await supabase.auth.getSession();

  console.log({ session });

  return res;
}

export const config = {
  matcher: ["/api/hello"],
};
```

And call one of the Supabase auth methods client-side, for example:

```
await supabase.auth.signInWithPassword({
      email: 'jon@supabase.com',
      password: 'password'
    });
```

Do you still get a null session?

**Comment by dijonmusters** [comment]
> Is there an ETA on the docs at all?

Aiming to get these live today 👍

**Comment by tiniscule** [comment]
> > How do i call inner nextjs api with session? `session` object in middleware returns `null`.
> 
> Very strange. This should only happen if you are not signed in.
> 
> If you make your middleware something like this:
> 
> ```
> import { createMiddlewareSupabaseClient } from "@supabase/auth-helpers-nextjs";
> import { NextResponse } from "next/server";
> import type { NextRequest } from "next/server";
> 
> export async function middleware(req: NextRequest) {
>   const res = NextResponse.next();
>   const supabase = createMiddlewareSupabaseClient({ req, res });
> 
>   const {
>     data: { session },
>   } = await supabase.auth.getSession();
> 
>   console.log({ session });
> 
>   return res;
> }
> 
> export const config = {
>   matcher: ["/api/hello"],
> };
> ```
> 
> And call one of the Supabase auth methods client-side, for example:
> 
> ```
> await supabase.auth.signInWithPassword({
>       email: 'jon@supabase.com',
>       password: 'password'
>     });
> ```
> 
> Do you still get a null session when navigating to your API route: `/api/hello`?

Yes - it happens on the first load after an auth.  The session doesn't get hydrated until the client interacts with it once - then a reload will show you it in the middleware.  But the first middleware hit after authentication always returns a null session for me.

**Comment by thorwebdev** [comment]
> But the first middleware hit after authentication always returns a null session for me.

Unfortunately that's to be expected for Oauth (redirect) based methods. The tokens are communicated via query fragments which cannot be accessed on the server. Therefore the client needs to render first, at which point supabase-js picks up the URL query fragments an creates the auth cookie. Once that has happened (which currently needs to happen client-side) the session will be accessible server-side via the auth cookie. 

You can find more details on this here: https://supabase.com/docs/guides/auth/server-side-rendering#understanding-the-authentication-flow

**Comment by dijonmusters** [comment]
Docs: https://github.com/supabase/supabase/pull/10396

**Comment by jensen** [comment]
> > Do you still get a null session when navigating to your API route: `/api/hello`?
> 
> Yes - it happens on the first load after an auth. The session doesn't get hydrated until the client interacts with it once - then a reload will show you it in the middleware. But the first middleware hit after authentication always returns a null session for me.

Not sure if this helps, but the way that I handle this is with an `/authenticated` route. https://github.com/jensen/supabase-nextjs/blob/main/app/(new)/authenticated/page.tsx

All it does is wait patiently for that session to be set. https://github.com/jensen/supabase-nextjs/blob/main/components/auth/check.tsx

I can tell the redirect to go to the `/authenticated` route using the supabase api. https://github.com/jensen/supabase-nextjs/blob/9cec28b5c48c16be7fd8487c586eadc90bfef401/components/auth/buttons/discord.tsx#L15

**Comment by tiniscule** [comment]
> Unfortunately that's to be expected for Oauth (redirect) based methods.

I think a recommendation on how to handle this officially would make sense then - as it basically means auth will feel broken using the default implementation example.  I wonder if there's another way to recognizing the first hit after auth to let it through to the original return path

**Comment by jensen** [comment]
> > Unfortunately that's to be expected for Oauth (redirect) based methods.
> 
> I think a recommendation on how to handle this officially would make sense then - as it basically means auth will feel broken using the default implementation example. I wonder if there's another way to recognize the first hit after auth to let it through to the original return path

I've spent dozens of hours working with Supabase auth, and I can't get around this with the hash based access token (when we are dealing with email confirmation or provide redirects). The only alternative I have tried was to alter GoTrue to send the tokens to my server route instead of redirecting to the client. This is why I have had to do what I showed above. 

When working with Remix, I went so far as to handle extracting the session from the URL directly without loading React etc. 

https://github.com/jensen/supabooked/blob/main/src/routes/authenticated.tsx

**Comment by rodzerjohn** [comment]
> > How do i call inner nextjs api with session? `session` object in middleware returns `null`.
> 
> Very strange. This should only happen if you are not signed in.
> 
> If you make your middleware something like this:
> 
> ```
> import { createMiddlewareSupabaseClient } from "@supabase/auth-helpers-nextjs";
> import { NextResponse } from "next/server";
> import type { NextRequest } from "next/server";
> 
> export async function middleware(req: NextRequest) {
>   const res = NextResponse.next();
>   const supabase = createMiddlewareSupabaseClient({ req, res });
> 
>   const {
>     data: { session },
>   } = await supabase.auth.getSession();
> 
>   console.log({ session });
> 
>   return res;
> }
> 
> export const config = {
>   matcher: ["/api/hello"],
> };
> ```
> 
> And call one of the Supabase auth methods client-side, for example:
> 
> ```
> await supabase.auth.signInWithPassword({
>       email: 'jon@supabase.com',
>       password: 'password'
>     });
> ```
> 
> Do you still get a null session when navigating to your API route: `/api/hello`?

I forgot to mention that I'm trying to access `/api/hello` from server side component using fetch API. In this case `session` returns `null` in middleware. How do i work with auth in this case?

Navigating to `/api/hello` works fine.

**Comment by NixBiks** [comment]
Wonderful docs @thorwebdev - especially going beyond and adding the data fetch + real-time updates pattern!

**Comment by noskovvkirill** [comment]
Having a trouble implementing it without redirect — i.e with phone based sign up. Page needs to be hard refreshed before user is authenticated. I set cookies manually and check them in middleware — it works for the server, but Supabase server client doesn't seem to catch it. 
**Middleware:**
  ```
 const supabase = createMiddlewareSupabaseClient({ req, res });

  const {
    data: { session }
  } = await supabase.auth.getSession();

let currentSession = session;

const refreshToken = req.cookies.get('my-refresh-token')?.value
const accessToken = req.cookies.get('my-access-token')?.value

  if (refreshToken && accessToken && refreshToken !== "" && accessToken !== "") {
    const { data, error } = await supabase.auth.setSession({ access_token: accessToken, refresh_token: refreshToken })
    if (error) {
      res.cookies.delete('my-refresh-token')
      res.cookies.delete('my-access-token')
    } else {
      if (data.session?.user && data.session?.access_token && data.session?.refresh_token) {
        currentSession = data.session
      }
    }
  }

```
And the page with the workaround 
```
 const supabase = createServerComponentSupabaseClient<Database>({
    headers,
    cookies
  });
  let { data: { user } } = await supabase.auth.getUser();
  let { data: { session } } = await supabase.auth.getSession();
  
  if (!user) {
    const cookies_ = cookies();
    const refreshToken = cookies_.get('my-refresh-token')?.value
    const accessToken = cookies_.get('my-access-token')?.value
    if (refreshToken && accessToken) {
      const { data, error } = await supabase.auth.setSession({ access_token: accessToken, refresh_token: refreshToken })
      if (error) {
        redirect('/')
      } else {
        session = data.session
        let { data: { user: user_ } } = await supabase.auth.getUser();
        user = user_
      }
    } else {
      redirect('/')
    }
  }
```

**Comment by hjaber** [comment]
> Having a trouble implementing it without redirect — i.e with phone based sign up. Page needs to be hard refreshed before user is authenticated. I set cookies manually and check them in middleware — it works for the server, but Supabase server client doesn't seem to catch it. 
> 
> **Middleware:**
>   ```
>  const supabase = createMiddlewareSupabaseClient({ req, res });
> 
>   const {
>     data: { session }
>   } = await supabase.auth.getSession();
> 
> let currentSession = session;
> 
> const refreshToken = req.cookies.get('my-refresh-token')?.value
> const accessToken = req.cookies.get('my-access-token')?.value
> 
>   if (refreshToken && accessToken && refreshToken !== "" && accessToken !== "") {
>     const { data, error } = await supabase.auth.setSession({ access_token: accessToken, refresh_token: refreshToken })
>     if (error) {
>       res.cookies.delete('my-refresh-token')
>       res.cookies.delete('my-access-token')
>     } else {
>       if (data.session?.user && data.session?.access_token && data.session?.refresh_token) {
>         currentSession = data.session
>       }
>     }
>   }
> 
> ```
> **And the page with the workaround:**
> ```
>  const supabase = createServerComponentSupabaseClient<Database>({
>     headers,
>     cookies
>   });
>   let { data: { user } } = await supabase.auth.getUser();
>   let { data: { session } } = await supabase.auth.getSession();
>   
>   // this works
>   if (!user) {
>     const cookies_ = cookies();
>     const refreshToken = cookies_.get('my-refresh-token')?.value
>     const accessToken = cookies_.get('my-access-token')?.value
>     if (refreshToken && accessToken) {
>       const { data, error } = await supabase.auth.setSession({ access_token: accessToken, refresh_token: refreshToken })
>       if (error) {
>         redirect('/')
>       } else {
>         session = data.session
>         let { data: { user: user_ } } = await supabase.auth.getUser();
>         user = user_
>       }
>     } else {
>       redirect('/')
>     }
>   }
> ```
> 
> Not sure what I'm doing wrong, but session that was set in the middleware isn't visible on the client if the page wasn't refreshed. I have to duplicate the logic in the component. 
> 
> @thorwebdev maybe you some ideas? 🥹

I have a similar issue relating to middleware rather than the server. 

When following the official docs for auth helpers on server components/app directory, the supabase server client immediately updates based on the session cookies and functions appropriately. 

But the middleware is out of sync with auth and requires the page to be refreshed before correctly identifying the current session. 

As the cookies are correctly being updated, I believe this to be a nextjs issue and not a supabase issue.

**Comment by tiniscule** [comment]
It's not really an "issue" on either side unfortunately - a design decision on the supabase side makes it impossible to detect the authenticated state prior to the client side page load.  The server never sees the hash params, so it has no ability to handle the authentication.  I took a stab at handling this here: https://usebasejump.com/blog/supabase-oauth-with-nextjs-middleware

**Comment by noskovvkirill** [comment]
@tiniscule interesting, now I understand the issue better. Thanks for sharing!

**Comment by hjaber** [comment]
> It's not really an "issue" on either side unfortunately - a design decision on the supabase side makes it impossible to detect the authenticated state prior to the client side page load. The server never sees the hash params, so it has no ability to handle the authentication. I took a stab at explaining and showing a workaround here: https://usebasejump.com/blog/supabase-oauth-with-nextjs-middleware

I thought this was only true for Oauth (redirect) based methods? 

I'm using `signInWithPassword()` and still unable to utilize middleware without a refresh while using the auth helpers server components/app implementation.

**Comment by goranefbl** [comment]
> > > Do you still get a null session when navigating to your API route: `/api/hello`?
> > 
> > 
> > Yes - it happens on the first load after an auth. The session doesn't get hydrated until the client interacts with it once - then a reload will show you it in the middleware. But the first middleware hit after authentication always returns a null session for me.
> 
> Not sure if this helps, but the way that I handle this is with an `/authenticated` route. https://github.com/jensen/supabase-nextjs/blob/main/app/(new)/authenticated/page.tsx
> 
> All it does is wait patiently for that session to be set. https://github.com/jensen/supabase-nextjs/blob/main/components/auth/check.tsx
> 
> I can tell the redirect to go to the `/authenticated` route using the supabase api. https://github.com/jensen/supabase-nextjs/blob/9cec28b5c48c16be7fd8487c586eadc90bfef401/components/auth/buttons/discord.tsx#L15

Thans. This works for SignIn, but not for SignUp if there is no email verification :/

**Comment by goranefbl** [comment]
> It's not really an "issue" on either side unfortunately - a design decision on the supabase side makes it impossible to detect the authenticated state prior to the client side page load. The server never sees the hash params, so it has no ability to handle the authentication. I took a stab at explaining and showing a workaround here: https://usebasejump.com/blog/supabase-oauth-with-nextjs-middleware

Thanks. This works for SignIn, but not for SignUp if there is no email verification. Any solution for that?

**Comment by csalmeida** [comment]
> > > Do you still get a null session when navigating to your API route: `/api/hello`?
> > 
> > 
> > Yes - it happens on the first load after an auth. The session doesn't get hydrated until the client interacts with it once - then a reload will show you it in the middleware. But the first middleware hit after authentication always returns a null session for me.
> 
> Not sure if this helps, but the way that I handle this is with an `/authenticated` route. https://github.com/jensen/supabase-nextjs/blob/main/app/(new)/authenticated/page.tsx
> 
> All it does is wait patiently for that session to be set. https://github.com/jensen/supabase-nextjs/blob/main/components/auth/check.tsx
> 
> I can tell the redirect to go to the `/authenticated` route using the supabase api. https://github.com/jensen/supabase-nextjs/blob/9cec28b5c48c16be7fd8487c586eadc90bfef401/components/auth/buttons/discord.tsx#L15

I have followed @jensen's approach but using `createBrowserClient` instead which is taken from the [NextJS Server Components example](https://github.com/supabase/auth-helpers/tree/main/examples/nextjs-server-components) but unfortunately, the session still returns `null` using OAuth.

Still investigating why this might be but thought I would leave my experience here in case anyone else is trying the same approach.

**Comment by trulysinclair** [comment]
@csalmeida for me I only had issues accidentally using both the supabasejs package and the auth helpers at the same time without noticing. Removing the main supabase package and making sure to only use auth helpers, everything worked perfectly fine.

**Comment by tonyxiao** [comment]
I'm getting this error when using the `createServerComponentSupabaseClient`. 
`Error: The "host" request header is not available`

Did anyone else run into this and have a workaround?
